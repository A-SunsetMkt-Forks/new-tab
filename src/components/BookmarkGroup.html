<div id="bookmarks">
  <div id="bookmarks-bar" ref:b>
    {#each __visible as __bookmark}
      <BookmarkItem __node="{__bookmark}" __lvl="{0}" />
    {/each}
  </div>

  {#if __overflow.children.length}
    <BookmarkItem __node="{__overflow}" __lvl="{0}" __left/>
  {/if}

  <BookmarkItem __node="{__bookmarksOther}" __lvl="{0}" __left/>
</div>

<svelte:window on:resize="__setVisibleEnd()"/>

<script>
  import BookmarkItem from './BookmarkItem.html';

  // Measure title text width using the 2d canvas measureText method as it's
  // faster than adding an element to the DOM -- but we need to manually
  // calculate any padding etc.

  const FONT = '18px sans-serif';
  const TITLE_MAX_WIDTH = 144;
  const FAVICON_WIDTH = 16;
  const FAVICON_PADDING = 9;
  const ITEM_PADDING = 9 + 9; // padding left + right

  // reusing the same canvas is much faster
  const canvas = document.createElement('canvas').getContext('2d');
  canvas.font = FONT;
  const ellipsisWidth = canvas.measureText('…').width;
  const titleMaxInnerWidth = TITLE_MAX_WIDTH - ellipsisWidth;

  export default {
    components: {
      BookmarkItem,
    },
    data: () => ({
      __bookmarksBar: [],
      __bookmarksOther: {},
      __end: 0,
    }),
    computed: {
      __visible: ({ __bookmarksBar, __end }) => __bookmarksBar.slice(0, __end),
      __overflow: ({ __bookmarksBar, __end }) => ({ title: '»', children: __bookmarksBar.slice(__end) }),
    },
    oncreate() {
      chrome.bookmarks.getTree((tree) => {
        this.set({
          __bookmarksBar: tree[0].children[0].children,
          __bookmarksOther: tree[0].children[1],
        });
      });
    },
    onstate({ changed, previous }) {
      if (previous === undefined) return;

      if (changed.__bookmarksBar) {
        this.__setVisibleEnd();
      }
    },
    methods: {
      __setVisibleEnd() {
        const { __bookmarksBar } = this.get();
        const bookmarksBarWidth = this.refs.b.offsetWidth - ITEM_PADDING;
        let __end = 0;
        let width = 0;

        __bookmarksBar.every((item, index) => {
          /* eslint-disable-next-line no-nested-ternary */
          const faviconWidth = item.children
            ? 0
            : item.title
              ? FAVICON_WIDTH + FAVICON_PADDING
              : FAVICON_WIDTH;
          const realTitleWidth = item.title ? canvas.measureText(item.title).width : 0;
          const titleWidth = realTitleWidth > titleMaxInnerWidth ? TITLE_MAX_WIDTH : realTitleWidth;
          const itemWidth = ITEM_PADDING + faviconWidth + titleWidth;
          const newWidth = width + itemWidth;

          if (newWidth < bookmarksBarWidth) {
            width = newWidth;
            __end = index + 1;
            return true;
          }
          return false;
        });

        this.set({ __end });
      },
    },
  };
</script>

<style type="text/postcss" global>
  #bookmarks-bar {
    width: 100%;
  }

  #bookmarks {
    position: fixed;
    top: 0;
    right: 0;
    left: 0;
    z-index: 2;
    display: flex;
    height: 41px;
    background: var(--c0);
    box-shadow: var(--s);
    backface-visibility: hidden; /* performance hack; force GPU */
  }
</style>
