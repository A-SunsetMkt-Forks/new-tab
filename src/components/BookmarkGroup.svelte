<script>
  import BookmarkNode from './BookmarkNode.svelte';

  // Get title text width using the 2d canvas measureText() method as it's
  // faster than adding an element to the DOM but we need to manually
  // calculate any padding etc.

  const FONT = '18px sans-serif';
  const TITLE_MAX_WIDTH = 144;
  const FAVICON_WIDTH = 16;
  const FAVICON_MARGIN = 6; // margin right
  const ITEM_PADDING = 13 + 13; // padding left + right

  // reuse the same canvas
  const canvas2d = new OffscreenCanvas(0, 0).getContext('2d');
  canvas2d.font = FONT;
  const ellipsisWidth = canvas2d.measureText('…').width;
  const titleMaxInnerWidth = TITLE_MAX_WIDTH - ellipsisWidth;

  // data
  let bookmarksBar = [];
  let bookmarksOther = { children: [] };
  let itemCount;
  let barEl;

  let visibleNodes;
  let overflowNodes;

  // computed properties
  $: visibleNodes = bookmarksBar.slice(0, itemCount);
  $: overflowNodes = ({ children: bookmarksBar.slice(itemCount), title: '»' });

  function handleResize() {
    const { length } = bookmarksBar;
    const otherBookmarksWidth = !bookmarksOther.children.length
      ? 0
      : canvas2d.measureText(bookmarksOther.title).width + ITEM_PADDING;
    // eslint-disable-next-line max-len
    const bookmarksBarWidth = barEl.offsetWidth - ITEM_PADDING - FAVICON_MARGIN - otherBookmarksWidth;
    let maxItems = 0;
    let width = 0;

    for (; maxItems < length; maxItems += 1) {
      // eslint-disable-next-line security/detect-object-injection
      const { children, title } = bookmarksBar[maxItems];
      const hasTitle = title !== undefined;
      // eslint-disable-next-line no-nested-ternary
      const faviconWidth = children !== undefined
        ? 0
        : hasTitle
          ? FAVICON_WIDTH + FAVICON_MARGIN
          : FAVICON_WIDTH;
      const realTitleWidth = hasTitle ? canvas2d.measureText(title).width : 0;
      const titleWidth = realTitleWidth > titleMaxInnerWidth
        ? TITLE_MAX_WIDTH
        : realTitleWidth;
      const itemWidth = ITEM_PADDING + faviconWidth + titleWidth;
      const nextWidth = width + itemWidth;

      if (nextWidth > bookmarksBarWidth) break;

      width = nextWidth;
    }

    itemCount = maxItems;
  }

  chrome.bookmarks.getTree((tree) => {
    bookmarksBar = tree[0].children[0].children;
    bookmarksOther = tree[0].children[1]; // eslint-disable-line prefer-destructuring

    handleResize();
  });

  window.addEventListener('resize', handleResize);
</script>

<div id="bookmarks" bind:this="{barEl}">
  {#each visibleNodes as _node}
    <BookmarkNode {_node} maxLen="{15}" lvl="{0}" />
  {/each}

  {#if overflowNodes.children.length}
    <BookmarkNode _node="{overflowNodes}" lvl="{0}" maxLen="{40}" endNode />
  {/if}

  {#if bookmarksOther.children.length}
    <BookmarkNode _node="{bookmarksOther}" lvl="{0}" maxLen="{40}" endNode />
  {/if}
</div>

<style type="text/postcss">
  :global(#bookmarks) {
    position: fixed;
    top: 0;
    right: 0;
    left: 0;
    z-index: 2;
    display: flex;
    height: 41px;
    background: var(--c0);
    box-shadow: var(--s);
    backface-visibility: hidden; /* performance hack; force GPU */
  }
</style>
