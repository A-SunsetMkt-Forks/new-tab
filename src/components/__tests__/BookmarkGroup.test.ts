import BookmarkGroup from '../BookmarkGroup.svelte';

describe('BookmarkGroup component', () => {
  it('renders correctly', () => {
    expect.assertions(1);
    const target = document.createElement('div');
    new BookmarkGroup({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });
});
