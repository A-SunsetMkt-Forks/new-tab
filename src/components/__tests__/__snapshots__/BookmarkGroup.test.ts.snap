// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkGroup component renders correctly 1`] = `"<div class=\\"bookmarks\\">  </div>"`;
